packages/excalidraw/change.ts

```diff
@@ -0,0 +1,1394 @@
+
+        const delta = Delta.create(
+          deleted,
+          inserted,
+          ElementsChange.stripIrrelevantProps,
+        );
+
+        removed.set(prevElement.id, delta);
+      }
+    }
+
+    for (const nextElement of nextElements.values()) {
+      const prevElement = prevElements.get(nextElement.id);
+
+      if (!prevElement) {
+        const deleted = { isDeleted: true } as ElementPartial;
+        const inserted = {
+          ...nextElement,
+          isDeleted: false,
+        } as ElementPartial;
+
+        const delta = Delta.create(
+          deleted,
+          inserted,
+          ElementsChange.stripIrrelevantProps,
+        );
+
+        added.set(nextElement.id, delta);
+
+        continue;
+      }
+
+      if (prevElement.versionNonce !== nextElement.versionNonce) {
+        const delta = Delta.calculate<ElementPartial>(
+          prevElement,
+          nextElement,
+          ElementsChange.stripIrrelevantProps,
+          ElementsChange.postProcess,
+        );
+
+        if (
+          // Making sure we don't get here some non-boolean values (i.e. undefined, null, etc.)
+          typeof prevElement.isDeleted === "boolean" &&
+          typeof nextElement.isDeleted === "boolean" &&
+          prevElement.isDeleted !== nextElement.isDeleted
+        ) {
+          // Notice that other props could have been updated as well
+          if (prevElement.isDeleted && !nextElement.isDeleted) {
+            added.set(nextElement.id, delta);
+          } else {
+            removed.set(nextElement.id, delta);
+          }
+
+          continue;
+        }
+
+        // Making sure there are at least some changes
+        if (!Delta.isEmpty(delta)) {
+          updated.set(nextElement.id, delta);
+        }
+      }
+    }
+
+    return ElementsChange.create(added, removed, updated);
+  }
+
+  public static empty() {
+    return ElementsChange.create(new Map(), new Map(), new Map());
+  }
+
+  public inverse(): ElementsChange {
+    const inverseInternal = (deltas: Map<string, Delta<ElementPartial>>) => {
+      const inversedDeltas = new Map<string, Delta<ElementPartial>>();
+
+      for (const [id, delta] of deltas.entries()) {
+        inversedDeltas.set(id, Delta.create(delta.inserted, delta.deleted));
+      }
+
+      return inversedDeltas;
+    };
+
+    const added = inverseInternal(this.added);
+    const removed = inverseInternal(this.removed);
+    const updated = inverseInternal(this.updated);
+
+    // Notice we inverse removed with added not to break the invariants
+    return ElementsChange.create(removed, added, updated);
+  }
+
+  public isEmpty(): boolean {
+    return (
+      this.added.size === 0 &&
+      this.removed.size === 0 &&
+      this.updated.size === 0
+    );
+  }
+
+  /**
+   * Update only `updated` delta/s based on the existing elements.
+   * In comparison, `added` and `removed` deltas are not updated as it would cause our invariants to break.
```

Comment:
Actually we should update added & removed, but not updating `isDeleted` flag, so the invariants will stand.
+ add test cases

Reply:
Done